ql/pricingengines/swaption/blackswaptionengine.hpp

```diff
@@ -222,8 +245,12 @@ namespace QuantLib {
 
         // using the discounting curve
         // swap.iborIndex() might be using a different forwarding curve
-        swap.setPricingEngine(ext::shared_ptr<PricingEngine>(new
-            DiscountingSwapEngine(discountCurve_, false)));
+        results_.valuationDate  = npvDate_;
+        if (npvDate_ == Date()) {
+            results_.valuationDate = discountCurve_->referenceDate();
```

Comment:
Instead of duplicating the calculation, you can retrieve `swap.valuationDate()` after setting the swap engine.